Add partial example of signing

steiza opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

For #136.

This isn't a full implementation of signing - it supports an on-disk keypair and a timestamp authority. It doesn't yet support Fulcio or Rekor. But I wanted to post a work-in-progress to get feedback on signing API generally.

We could choose to land this functionality and have a follow-on PR that implements Fulcio and Rekor support (and enables conformance testing on signing) - smaller reviews are nice! Or I could continue to update this PR with more functionality.

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

NONE - for now, until at least Fulcio is supported

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

NONE

steiza committed:
Adding example of signing.

Supports on-disk keypair and timestamp authority. Does not (yet) support
Fulcio or Rekor.

Signed-off-by: Zach Steindler <steiza@github.com>

Hayden-IO commented on pkg/sign/witness.go:
We need a different name for this concept, this will conflict with transparency log witnessing.

Actually, per the comment below, I would just call this timestamping.

Hayden-IO commented on pkg/sign/witness.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+// TODO: implement
+// type Rekor struct {
```
I mentioned this just now on the issue, but I don't think we should conflate timestamping and logging. These are two different concepts. It just so happens that the log can provide a timestamp, and it can be trusted by the client policy if they want to trust the log.

I would rather have two separate steps, one for timestamping (which could fetch a timestamp from a TSA or a Roughtime server or something else in the future), and one for logging (calling out to N logs). It's up to the client to determine if it trusts the timestamp returned from the last step.

steiza replied:
👍 When we get to it, we'll put Rekor logging in a separate file in the `sign` directory.

Hayden-IO commented on pkg/sign/witness.go:
```diff
@@ -0,0 +1,113 @@
+		return err
+	}
+
+	client, err := tsaclient.GetTimestampClient(ta.baseURL, tsaclient.WithUserAgent("sigstore-go"), tsaclient.WithContentType(tsaclient.TimestampQueryMediaType))
```
Can we include version in the user agent (if that's possible to dynamically fetch?)?

steiza replied:
For now I added this as an optional parameter to `TimestampAuthorityOptions`, populated by `git describe --tags` in the `Makefile`. @codysoyland what do you think about how/if we should track the version of `sigstore-go` in the codebase?

Hayden-IO commented on pkg/sign/witness.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+type TimestampAuthority struct {
+	baseURL string
```
We should include a context or common parameters like timeout to initialize the client.

Hayden-IO commented on pkg/sign/signer.go:
```diff
@@ -0,0 +1,95 @@
+	publicKeyHint []byte
+}
+
+func SignerKeypair(s crypto.Signer, ha protocommon.HashAlgorithm, pkh []byte) Keypair {
```
Do we want to make an options struct to initialize the signer? Then if we need to expand the options in the future, it won't require a breaking API change.

Hayden-IO commented on pkg/sign/signer.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+func SignerKeypair(s crypto.Signer, ha protocommon.HashAlgorithm, pkh []byte) Keypair {
+	return Keypair{signer: s, hashAlgorithm: ha, publicKeyHint: pkh}
```
For `publicKeyHint`, can we make this an optional parameter, and if it's not given, derive it ourselves (`SHA-256(DER(SPKI(pub key)))`)? Most clients won't be opinionated on the key hint I suspect.

steiza replied:
I'm not sure I produced the hint in exactly the way you suggested, but we do now generate a hint string if one is not provided.

Hayden-IO replied:
Looks great! That's equivalent, DER(SPKI()) is the same as a marshaled PKIX key.

Hayden-IO reviewed (changes requested):
+100 to small PRs! This is much easier to review! Let's go with incremental PRs, and I suspect most of the discussion will be in the first couple PRs as we figure out the interface.

steiza committed:
Feedback from code review

- Add signer options, make key hint optional, and add tests
- Rename "witness" to "timestamping", add client timeout and
  user agent

Signed-off-by: Zach Steindler <steiza@github.com>

Hayden-IO commented on pkg/sign/signer.go:
```diff
@@ -0,0 +1,113 @@
+	"crypto"
+	"crypto/rand"
+	"crypto/sha256"
+	_ "crypto/sha512" // if user chooses SHA2-384 or SHA2-256 for hash
```
```suggestion
	_ "crypto/sha512" // if user chooses SHA2-384 or SHA2-512 for hash
```

Hayden-IO reviewed (dismissed):
Looks great!

steiza committed:
Update pkg/sign/signer.go

Co-authored-by: Hayden B <hblauzvern@google.com>
Signed-off-by: Zach Steindler <steiza@github.com>
